internal/alpini/router/alpini-router-api/src/main/java/com/linkedin/alpini/router/api/Scatter.java

```diff
@@ -16,11 +16,11 @@
  */
 public class Scatter<H, P extends ResourcePath<K>, K> {
   /** Requests for servers that are available. */
-  private final Collection<ScatterGatherRequest<H, K>> _onlineRequests = new ArrayList<>();
+  private final Collection<ScatterGatherRequest<H, K>> _onlineRequests = new ConcurrentLinkedQueue<>();
   private final Collection<ScatterGatherRequest<H, K>> _unmodifiableOnlineRequests =
       Collections.unmodifiableCollection(_onlineRequests);
   /** Requests for servers that are offline. */
-  private final Collection<ScatterGatherRequest<H, K>> _offlineRequests = new ArrayList<>();
+  private final Collection<ScatterGatherRequest<H, K>> _offlineRequests = new ConcurrentLinkedQueue<>();
```

Comment:
I am not super thrilled about changing these list implementations... IME, `ArrayList` is quite cheap (in terms of memory/garbage as well as CPU) and therefore pretty hard to beat. A linked list will add a wrapper object around each element, and will also make `getOnlineRequestCount()` slower (O(N) rather than O(1)).

So although it is right to make the new behaviour configurable, this PR actually risks changing the performance profile of the code even when the new config is off.

I think, instead of instantiating these variables at their definition-site, we should do it in the constructor, and the code which calls the constructor should pass the right implementation based on the new config. Thankfully, there are very few code paths calling the constructor, so this should be easy.

Reply:
Makes sense and fixed.